File: src/Cli/Journal/Digest.php
<?php

declare(strict_types=1);

namespace JesseGall\CodeCommandments\Cli\Journal;

use JesseGall\CodeCommandments\Cli\Text;

/**
 * Chooses what is worth reading out of a stretch of conversation. A transcript is far too long to hand
 * back whole and a summary of it is what lost the decisions in the first place, so this SELECTS: the
 * user's own words untouched, enough either side of them to know what they were answering, and — through
 * the long stretches where the agent worked alone — only the messages that said they carried something.
 */
final class Digest
{
    /**
     * How many of the agent's messages before a prompt are kept. A bare "yes please" means nothing without
     * the thing it answered, and the answer is usually the message right before it.
     */
    private const int LEADING = 2;

    /**
     * How many after it — what the agent said it would do about what was just said.
     */
    private const int TRAILING = 1;

    /**
     * How many messages may be skipped before the gap is worth mentioning. Below this it is a pause;
     * above it, the agent was working alone and the reader should know.
     */
    private const int GAP = 3;

    /**
     * How far a message is indented from the speaker's label — the width of `USER  ` and `  me  `, so a
     * wrapped line lands under the words rather than under the name.
     */
    private const int GUTTER = 6;

    /**
     * @param  list<Line>  $lines
     */
    public function __construct(private readonly array $lines) {}

    /**
     * The lines of THIS digest that somebody said.
     *
     * @return list<Line>
     */
    private function spoken(): array
    {
        return self::spokenIn($this->lines);
    }

    /**
     * The lines of $lines that somebody SAID, with a prompt the user kept typing counted once. The harness
     * records a message as it is sent and again when more of it follows, so the same words arrive twice
     * with the second carrying the rest; the earlier is a prefix of the later, which identifies it.
     *
     * @param  list<Line>  $lines
     * @return list<Line>
     */
    public static function spokenIn(array $lines): array
    {
        $spoken = array_values(array_filter($lines, fn (Line $line) => $line->isSpeech() && $line->text !== ''));
        $kept = [];

        foreach ($spoken as $at => $line) {
            $next = $spoken[$at + 1] ?? null;

            if ($next !== null && $line->isPrompt() && $next->isPrompt() && str_starts_with($next->text, $line->text)) {
                continue;
            }

            $kept[] = $line;
        }

        return $kept;
    }

    /**
     * The chosen lines, in order.
     *
     * @return list<Line>
     */
    public function selected(): array
    {
        $speech = $this->spoken();
        $keep = [];

        foreach ($speech as $at => $line) {
            if ($this->isWorthKeeping($speech, $at, $line)) {
                $keep[$at] = $line;
            }
        }

        return array_values($keep);
    }

    /**
     * The digest as it reads, with the pinned facts first and the stretches the agent worked alone through
     * marked rather than silently dropped.
     */
    public function render(): string
    {
        $speech = $this->spoken();
        $written = [];
        $previous = -1;

        foreach ($speech as $at => $line) {
            if (! $this->isWorthKeeping($speech, $at, $line)) {
                continue;
            }

            $skipped = $at - $previous - 1;

            if ($previous >= 0 && $skipped >= self::GAP) {
                $written[] = "      ⋯ {$skipped} messages ⋯";
            }

            $written[] = $this->line($line);
            $previous = $at;
        }

        return implode("\n", [...$this->pinned($speech), ...$written]);
    }

    /**
     * Is this line one the reader needs? The user's own words always; the agent's when they said what they
     * carried, or when they sit close enough to a prompt to be its context.
     *
     * @param  list<Line>  $speech
     */
    private function isWorthKeeping(array $speech, int $at, Line $line): bool
    {
        if ($line->isPrompt()) {
            return true;
        }

        if ($line->tag()->isSome()) {
            return true;
        }

        return $this->isNearAPrompt($speech, $at);
    }

    /**
     * Does a prompt stand close enough behind or ahead of $at for this line to be its context?
     *
     * @param  list<Line>  $speech
     */
    private function isNearAPrompt(array $speech, int $at): bool
    {
        for ($near = $at - self::TRAILING; $near <= $at + self::LEADING; $near++) {
            if (($speech[$near] ?? null)?->isPrompt() === true) {
                return true;
            }
        }

        return false;
    }

    /**
     * The facts marked to outlive every compaction, gathered at the top where they cannot be missed.
     *
     * @param  list<Line>  $speech
     * @return list<string>
     */
    private function pinned(array $speech): array
    {
        $pinned = array_filter($speech, fn (Line $line) => $line->tag()->isSomeAnd(fn (Tag $tag) => $tag->isPinned()));

        if ($pinned === []) {
            return [];
        }

        return [Text::heading('pinned'), '', ...array_map($this->line(...), $pinned), ''];
    }

    /**
     * One line as it reads — the user in their own words, the agent indented behind them.
     */
    private function line(Line $line): string
    {
        return $line->isPrompt()
            ? 'USER  ' . $this->wrapped($line->text)
            : '  me  ' . $this->wrapped($line->text);
    }

    /**
     * $text laid out under the speaker's label — wrapped to the terminal, every later line indented to sit
     * beneath the first. A paragraph the window folds where it likes is a wall; nothing is CUT, since the
     * user's words are the whole reason for reading.
     */
    private function wrapped(string $text): string
    {
        return Text::reflow($text, self::GUTTER);
    }
}

File: .journal/digest.py
"""What is worth reading back, out of everything that was said.

A digest keeps the user's own words UNTOUCHED and enough either side of them to know what
they were answering. A bare "yes please" is meaningless without the thing it agreed to,
which is why a prompt is never shown alone.

Everything else is kept only when the agent said it CARRIED something. A `[!reply]` obeyed
the rule and carries nothing, so it is counted in an elision rather than shown — the same
as an untagged one. A quiet stretch reads back as `⋯ 41 messages ⋯`, which is its worth.
"""
from __future__ import annotations

import tags
import transcript

CONTEXT = 2  # messages either side of a prompt that are its context


def _near_prompt(lines, i: int, prompts: set[int]) -> bool:
    """Does a prompt sit close enough behind or ahead for this line to be its context?"""
    return any(abs(i - p) <= CONTEXT for p in prompts)


def select(lines) -> list:
    """The lines a reader needs, in order.

    The user's words ALWAYS; the agent's when they carried a tag, or when they sit close
    enough to a prompt to be what it answered or what answered it.

    A message with no TEXT is not context: a turn that only called a tool said nothing, and
    keeping it spends a line of the digest to show a blank.
    """
    # ONLY A FILING UNIT CAN BE KEPT FOR PROXIMITY. Nearness to a prompt is a weak reason
    # — it keeps a line for where it sat, not for what it said — and applied to every text
    # block it dragged in the scaffolding: "Now wiring it into the CLI", "Let me check the
    # other file". The hook already stopped treating those as messages. This is the same
    # rule, read from the same place, so the two halves of the system cannot drift again.
    #
    # A TAG STILL WINS ON ITS OWN. A line that declared what it carried is kept whether or
    # not it ended its turn: the agent said it was worth reading back, and the digest has
    # no business overruling that with a structural guess.
    units = transcript.filing_units(lines)
    spoken = [l for l in lines if l.spoken and (l.text or "").strip()]
    prompts = {i for i, l in enumerate(spoken) if l.kind == "human"}
    # THE MESSAGE THE USER REACTED TO, AND THE ONE THAT ANSWERED THEM. A prompt is nearly
    # always a reaction: a correction, a steer, a "yes, do that". Shown alone it is a reply
    # to nothing. Proximity mostly keeps these, but a hook hold between the answer and the
    # prompt pushes the answer out of range — so the last filed message before each prompt
    # and the first after it are kept by name, whatever the distance.
    around: set[int] = set()
    unit_idx = [i for i, l in enumerate(spoken) if l.n in units]
    for p in prompts:
        before = [i for i in unit_idx if i < p]
        after = [i for i in unit_idx if i > p]
        if before:
            around.add(before[-1])
        if after:
            around.add(after[0])
    keep = []
    for i, line in enumerate(spoken):
        if (
            line.kind == "human"
            or tags.carried(line.text)
            or i in around
            or (line.n in units and _near_prompt(spoken, i, prompts))
        ):
            keep.append(line)
    return keep


def render(lines, *, elide: bool = True) -> str:
    """The digest as it reads: the user in their own words, the agent indented behind them.

    A dropped stretch is COUNTED rather than hidden. A reader who cannot see that forty
    messages were skipped cannot tell a quiet stretch from a missing one.
    """
    keep = select(lines)
    kept = {l.n for l in keep}
    out: list[str] = []
    skipped = 0
    for line in (l for l in lines if l.spoken and (l.text or "").strip()):
        if line.n not in kept:
            skipped += 1
            continue
        if skipped and elide:
            out.append(f"      ⋯ {skipped} message(s) ⋯")
        skipped = 0
        out.append(_one(line))
    if skipped and elide:
        out.append(f"      ⋯ {skipped} message(s) ⋯")
    return "\n".join(out)


def _one(line) -> str:
    found = tags.found(line.text)
    body = tags.strip(line.text) if found else (line.text or "").strip()
    body = " ".join(body.split())
    if line.kind == "human":
        return f"\n{line.n:>5}  ▸ USER: {body}"
    mark = f"[{'/'.join('!' + t for t in found)}] " if found else ""
    return f"{line.n:>5}    {mark}{body[:600]}"


def users_only(lines) -> str:
    """Only the user's own words, in full, never trimmed.

    The one tier that is never summarised: what somebody TOLD the agent is the half a
    compaction is most likely to drop and the half nothing else can re-derive.
    """
    out = []
    for line in lines:
        if line.kind == "human":
            out.append(f"\n{line.n:>5}  {line.ts[:19]}\n{(line.text or '').strip()}")
    return "\n".join(out)
